dm/pkg/schema/tracker.go

```diff
@@ -329,3 +359,225 @@ func (tr *Tracker) CreateTableIfNotExists(table *filter.Table, ti *model.TableIn
+	stmtNode, err := tr.dsTracker.stmtParser.ParseOneStmt(createStr, "", "")
+	if err != nil {
+		return nil, dterror.ErrSchemaTrackerInvalidCreateTableStmt.Delegate(err, createStr)
+	}
+
+	ti, err := ddl.MockTableInfo(mock.NewContext(), stmtNode.(*ast.CreateTableStmt), mockTableID)
+	if err != nil {
+		return nil, dterror.ErrSchemaTrackerCannotMockDownstreamTable.Delegate(err, createStr)
+	}
+	return ti, nil
+}
+
+// initDownStreamTrackerParser init downstream tracker parser by default sql_mode.
+func (tr *Tracker) initDownStreamSQLModeAndParser(tctx *tcontext.Context) error {
+	setSQLMode := fmt.Sprintf("SET SESSION SQL_MODE = '%s'", mysql.DefaultSQLMode)
+	_, err := tr.dsTracker.downstreamConn.ExecuteSQL(tctx, []string{setSQLMode})
+	if err != nil {
+		return dterror.ErrSchemaTrackerCannotSetDownstreamSQLMode.Delegate(err, mysql.DefaultSQLMode)
+	}
+	stmtParser, err := utils.GetParserFromSQLModeStr(mysql.DefaultSQLMode)
+	if err != nil {
+		return dterror.ErrSchemaTrackerCannotInitDownstreamParser.Delegate(err, mysql.DefaultSQLMode)
+	}
+	tr.dsTracker.stmtParser = stmtParser
+	return nil
+}
+
+// getDownStreamTi constructs downstreamTable index cache by tableinfo.
+func getDownStreamTi(ti *model.TableInfo, originTi *model.TableInfo) *downstreamTableInfo {
+	var (
+		indexCache       *model.IndexInfo
+		availableUKCache = make([]*model.IndexInfo, 0, len(ti.Indices))
+		hasPk            = false
+	)
+
+	// func for check not null constraint
+	fn := func(i int) bool {
+		return mysql.HasNotNullFlag(ti.Columns[i].Flag)
+	}
+
+	for _, idx := range ti.Indices {
+		if !idx.Primary && !idx.Unique {
+			continue
+		}
+		indexRedict := redirectIndexKeys(idx, originTi)
+		if indexRedict == nil {
+			continue
+		}
+		if idx.Primary {
+			indexCache = indexRedict
+			hasPk = true
+		} else if idx.Unique {
+			// second check not null unique key
+			if !hasPk && isSpecifiedIndexColumn(idx, fn) {
+				indexCache = indexRedict
+			} else {
+				availableUKCache = append(availableUKCache, indexRedict)
+			}
+		}
+	}
+
+	// handle pk exceptional case.
+	// e.g. "create table t(a int primary key, b int)".
+	if !hasPk {
+		exPk := redirectIndexKeys(handlePkExCase(ti), originTi)
+		if exPk != nil {
+			indexCache = exPk
+		}
+	}
+
+	return &downstreamTableInfo{
+		tableInfo:        ti,
+		indexCache:       indexCache,
+		availableUKCache: availableUKCache,
+	}
+}
+
+// redirectIndexKeys redirect index's columns offset in origin tableinfo.
+func redirectIndexKeys(index *model.IndexInfo, originTi *model.TableInfo) *model.IndexInfo {
+	if index == nil || originTi == nil {
+		return nil
+	}
+
+	columns := make([]*model.IndexColumn, 0, len(index.Columns))
+	for _, key := range index.Columns {
+		if originColumn := model.FindColumnInfo(originTi.Columns, key.Name.L); originColumn != nil {
+			column := &model.IndexColumn{
+				Name:   key.Name,
+				Offset: originColumn.Offset,
+				Length: key.Length,
+			}
+			columns = append(columns, column)
+		}
```

Comment:
if any column is not found, we can return earlier.

Reply:
yes, it will more concise, i will fix.